### PR TITLE
Fix if loop related lint failures in the deepcopy-gen

### PR DESCRIPTION
**Pull request by spzala**
The changes in this patch takes care of lint failures due to unreachable
else statements. For example one of the failures is,
/zz_generated.deepcopy.go:116:10: if block ends with a return statement,
so drop this else and outdent its block

### CODE REVIEW

File: examples/deepcopy-gen/generators/deepcopy.go [hunk 1]
```diff
@@ -523,10 +523,10 @@ func (g *genDeepCopy) doMap(t *types.Type, sw *generator.SnippetWriter) {
 				sw.Do("}\n", nil)
 				sw.Do("(*out)[key] = *newVal\n", nil)
 			} else {
-				sw.Do("if newVal, err := c.DeepCopy(&val); err != nil {\n", nil)
-				sw.Do("return err\n", nil)
-				sw.Do("} else {\n", nil)
+				sw.Do("if newVal, err := c.DeepCopy(&val); err == nil {\n", nil)
```
Comment (thockin) [inline examples/deepcopy-gen/generators/deepcopy.go]: It's much more idiomatic to test err != nil - why change that?  I had a hard time realizing what you changed because my eyes just skip the pattern.  I'd rather see a test for != nil and a `return nil` in the success path

Comment (spzala) [inline examples/deepcopy-gen/generators/deepcopy.go]: @thockin Hi Tim, lol, sorry and I agree. Honestly I didn't like it myself too but I notice use of == nil in few other places and I was getting locally scoped error with a different solution I tried few weeks back (part of an old PR which didn't go further because of underscore related changes). But seems like a possible better solution is working now, I will update the changes. Thanks!

**Comment by thockin** [comment]
This is an annoying and stupid lint error.  The `if x := func(); x { ... } else { ... }` is idiomatic and should be encouranged when x is not used outside the condition blocks.  

sigh

**Comment by thockin** [comment]
change LGTM but fails travis

**Comment by thockin** [comment]
restarted travis - 1/3 failed

**Comment by spzala** [comment]
@thockin I know :( and thanks much!! It seems like no straightforward error with Travis? So I guess may be one more restart on a weekday?

**Comment by thockin** [comment]
It is failing against Go tip, consistently.  I am on mobile, so can't investigate.

```
--- FAIL: TestSnippetWriter (0.00s)
	snippet_writer_test.go:85: Expected "snippet_writer_test.go:78" but didn't find it in "template: /home/travis/gopath/src/k8s.io/gengo/generator/snippet_writer_test.go:77:1: unclosed action"
FAIL
FAIL	k8s.io/gengo/generator	0.006s
```

**Comment by spzala** [comment]
@thockin Thanks Tim and no rush!! I noticed that it was the same failure even in previous patch with no code change (i.e. only  a comment was added https://github.com/kubernetes/gengo/pull/45 ). Seems strange :(

**Comment by thockin** [comment]
It looks like something change in upstream Go wrt line numbers.

Can you fold in a change (separate commit) to `generator/snippet_writer_test.go` to remove the literal line number? and see if that passes?

**Comment by spzala** [comment]
@thockin Hi Tim, very nice, that seems did the trick .. I created this one https://github.com/kubernetes/gengo/pull/48 and Travis's happy. Thanks!!

**Comment by thockin** [comment]
rebase and ping me

**Comment by spzala** [comment]
@thockin Hi Tim, I tried pinging you on slack kube-dev couple times but without much luck of talking to you :-) Just an update that this PR is passing Travis. Thanks!!

**Comment by thockin** [comment]
I just restarted travis - I added tests that this will have to update.
Resync to master and look at examples/deepcopy/output_tests

**Comment by thockin** [comment]
Specifically, run `make test` in that dir.

**Comment by spzala** [comment]
Thanks Tim @thockin

**Comment by spzala** [comment]
@thockin is there a way I can restart Travis job by adding a comment or something like that? (I did rebase and updated PR but seems like latest Travis failure is due to incomplete build job and probably need a restart) Thanks!

**Comment by thockin** [comment]
like that

**Comment by spzala** [comment]
Ahhh.. interesting.. Close the PR and ReOpen. Different but that's cool :-) Thanks @thockin !!

**Comment by spzala** [comment]
@thockin Yay :-) it passed now.

**Comment by thockin** [comment]
Curiously, no test cases exercize this.  I'm going to merge it because that's not your fault, but if you want to make me extra happy, maybe you could throw together super-minimal output_test cases that expose this change?

**Comment by spzala** [comment]
@thockin :-) Thanks Tim! And yes sure, I would work on it. I am traveling on business for a week and preparing for my speaking session etc but I believe there is no super rush with it right? I will be working on soon I can.  Thanks much!!

**Comment by thockin** [comment]
Also it should be SUPER EASY.  Copy on of the directories, edit doc.g to
have a type that highlights this case, run `make test`

**Comment by spzala** [comment]
Awesome, thanks @thockin !!

File: examples/deepcopy-gen/generators/deepcopy.go [hunk 2]
```diff
@@ -523,11 +523,12 @@ func (g *genDeepCopy) doMap(t *types.Type, sw *generator.SnippetWriter) {
 				sw.Do("}\n", nil)
 				sw.Do("(*out)[key] = *newVal\n", nil)
 			} else {
-				sw.Do("if newVal, err := c.DeepCopy(&val); err != nil {\n", nil)
+				sw.Do("newVal, err := c.DeepCopy(&val)\n", nil)
```
Comment (liggitt) [inline examples/deepcopy-gen/generators/deepcopy.go]: this produces code that will not compile

see https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/45797/pull-kubernetes-bazel/20933/

```
func DeepCopy_apiextensions_CustomResource(in interface{}, out interface{}, c *conversion.Cloner) error {
	{
		in := in.(*CustomResource)
		out := out.(*CustomResource)
		*out = *in
		newVal, err := c.DeepCopy(&in.ObjectMeta)
		if err != nil {
			return err
		}
		out.ObjectMeta = *newVal.(*v1.ObjectMeta)

		newVal, err := c.DeepCopy(&in.Spec)
		if err != nil {
			return err
		}
		out.Spec = *newVal.(*CustomResourceSpec)

		newVal, err := c.DeepCopy(&in.Status)
		if err != nil {
			return err
		}
		out.Status = *newVal.(*CustomResourceStatus)

		return nil
	}
}
```

Comment (liggitt) [inline examples/deepcopy-gen/generators/deepcopy.go]: opened https://github.com/kubernetes/gengo/pull/56 to revert so we can pull master (we need to bump as a prereq of https://github.com/kubernetes/kubernetes/pull/45294)... can revisit this if we decide we want to jump through scoping or variable name hoops to avoid this lint error (I personally don't think it's worth it)

Comment (spzala) [inline examples/deepcopy-gen/generators/deepcopy.go]: @liggitt thanks for fixing it Jordan. My bad, in my manual verification of generated code and build I apparently didn't run into it. I guess Gengo Travis passed because the method with nil return is not in Gengo but K8s. @thockin was quick in seeing a need for tests when lint related these changes were merged and kindly created new tests but those tests still missed few scenarios. I have also create a follow up PR (https://github.com/kubernetes/gengo/pull/51) to add new tests.  I value your opinion on spending time for lint may not worth :-) but if we can fix existing code on lint and going forward have lint complied code then that will reduce skipping lint check and produce more clean code. Thanks!!

Comment (liggitt) [inline examples/deepcopy-gen/generators/deepcopy.go]: lint is a moving target, and (as in this case) sometimes encourages non-idiomatic code. I don't think jumping through hoops to silence it is a good i dea.

Comment (liggitt) [inline examples/deepcopy-gen/generators/deepcopy.go]: > I have also create a follow up PR (#51) to add new tests.

I don't think that test exercises this case, either... this error was caused by multiple fields in the deep-copied type resulting in newVal being redeclared multiple times in the same scope.

Comment (spzala) [inline examples/deepcopy-gen/generators/deepcopy.go]: @liggitt about lint agree as long as we try to catch it manually but at the same time I guess what can be worth is to automate test process against lint (latest) as part of the build (something similar to what's done with PEP 8 in projects like OpenStack), that should be a real solution.

Comment (spzala) [inline examples/deepcopy-gen/generators/deepcopy.go]: @liggitt  for #51 do you think whatever is added in the PR are valuable tests for those specific scenarios and good be have? Thanks!
